packages/contracts-bedrock/src/cannon/PreimageOracle.sol

```diff
@@ -3,18 +3,77 @@ pragma solidity 0.8.15;
+    //                         Constants                          //
+    ////////////////////////////////////////////////////////////////
+
+    /// @notice The depth of the keccak256 merkle tree. Supports up to 65,536 keccak blocks, or ~8.91MB preimages.
+    uint256 public constant KECCAK_TREE_DEPTH = 16;
+    /// @notice The maximum number of keccak blocks that can fit into the merkle tree.
+    uint256 public constant MAX_LEAF_COUNT = 2 ** KECCAK_TREE_DEPTH - 1;
+    /// @notice The duration of the large preimage proposal challenge period.
+    uint256 public constant CHALLENGE_PERIOD = 1 days;
+
+    ////////////////////////////////////////////////////////////////
+    //                 Authorized Preimage Parts                  //
+    ////////////////////////////////////////////////////////////////
+
     /// @notice Mapping of pre-image keys to pre-image lengths.
     mapping(bytes32 => uint256) public preimageLengths;
     /// @notice Mapping of pre-image keys to pre-image parts.
     mapping(bytes32 => mapping(uint256 => bytes32)) public preimageParts;
     /// @notice Mapping of pre-image keys to pre-image part offsets.
     mapping(bytes32 => mapping(uint256 => bool)) public preimagePartOk;
 
+    ////////////////////////////////////////////////////////////////
+    //                  Large Preimage Proposals                  //
+    ////////////////////////////////////////////////////////////////
+
+    /// @notice A raw leaf of the large preimage proposal merkle tree.
+    struct Leaf {
+        /// @notice The input absorbed for the block, exactly 136 bytes.
+        bytes input;
+        /// @notice The index of the block in the absorbtion process.
+        uint256 index;
+        /// @notice The hash of the internal state after absorbing the input.
+        bytes32 stateCommitment;
+    }
+
+    /// @notice Static padding hashes. These values are persisted in storage, but are entirely immutable
+    ///         after the constructor's execution.
+    bytes32[KECCAK_TREE_DEPTH] public zeroHashes;
+    /// @notice Mapping of claimants to proposal UUIDs to the current branch path of the merkleization process.
+    mapping(address => mapping(uint256 => bytes32[KECCAK_TREE_DEPTH])) public proposalBranches;
+    /// @notice Mapping of claimants to proposal UUIDs to the timestamp of creation of the proposal as well as the
+    /// challenged status.
+    mapping(address => mapping(uint256 => LPPMetaData)) public proposalMetadata;
+    /// @notice Mapping of claimants to proposal UUIDs to the preimage part picked up during the absorbtion process.
+    mapping(address => mapping(uint256 => bytes32)) public proposalParts;
+
+    ////////////////////////////////////////////////////////////////
+    //                        Constructor                         //
+    ////////////////////////////////////////////////////////////////
+
+    constructor() {
+        // Compute hashes in empty sparse Merkle tree
```

Comment:
I'd add a comment here to make it explicit that the zero indexed `zeroHash` should be zero, since this loop effectively starts at 1.